### PR TITLE
[ALG-122] Simplify SWE CNE creator

### PR DESCRIPTION
**Pull request by MartinBelthle**
**Please check if the PR fulfills these requirements** 
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** 

No

**What kind of change does this PR introduce?** 

Refactoring

**What is the current behavior?** 

We need both RaoResult and AngleMonitoringResult to create the SweCne

**What is the new behavior?**

Now we only need the RaoResult

**Does this PR introduce a breaking change?** 

If the method `exportCne()` in `SweCneExporter` class is used by the SWE process, indeed it's a breaking change.
The process will have to create an object `RaoResultWithAngleMonitoring` with both `RaoResult` and `AngleMonitoringResult` to give it to the method.

**Other information**:

There is a cucumber branch: `alg-122-simplify-swe-cne-creator` with the API changes described just above.
Every test passed

There's also a PR with updated doc : https://github.com/farao-community/farao-community.github.io/pull/93

### CODE REVIEW

File: data/result-exporter/swe-cne-exporter/src/main/java/com/farao_community/farao/data/swe_cne_exporter/SweCneHelper.java [hunk 1]
```diff
@@ -29,9 +30,9 @@ public class SweCneHelper extends CneHelper {
     private final AngleMonitoringResult angleMonitoringResult;
     private Map<Contingency, Boolean> contingencyFailureMap = new HashMap<>();
 
-    public SweCneHelper(Crac crac, Network network, RaoResult raoResult, AngleMonitoringResult angleMonitoringResult, RaoParameters raoParameters, CneExporterParameters exporterParameters) {
+    public SweCneHelper(Crac crac, Network network, RaoResult raoResult, RaoParameters raoParameters, CneExporterParameters exporterParameters) {
         super(crac, network, raoResult, raoParameters, exporterParameters);
-        this.angleMonitoringResult = angleMonitoringResult;
+        this.angleMonitoringResult = ((RaoResultWithAngleMonitoring) raoResult).getAngleMonitoringResult();
```
Comment (pet-mit) [inline data/result-exporter/swe-cne-exporter/src/main/java/com/farao_community/farao/data/swe_cne_exporter/SweCneHelper.java:35]: you shouldn't need angleMonitoringResult anymore, angle values can be accessed in RaoResult

File: data/result-exporter/swe-cne-exporter/src/test/resources/SweCNEDivergentAngleMonitoring_Z01.xml
Comment (pet-mit) [inline data/result-exporter/swe-cne-exporter/src/test/resources/SweCNEDivergentAngleMonitoring_Z01.xml:1]: why did this change ?

File: monitoring/angle-monitoring/src/main/java/com/farao_community/farao/monitoring/angle_monitoring/RaoResultWithAngleMonitoring.java [hunk 2]
```diff
@@ -35,17 +36,23 @@ public RaoResultWithAngleMonitoring(RaoResult raoResult, AngleMonitoringResult a
...
+    public AngleMonitoringResult getAngleMonitoringResult() {
+        return angleMonitoringResult;
+    }
+
     @Override
     public ComputationStatus getComputationStatus() {
-        if (angleMonitoringResult.isDivergent() || angleMonitoringResult.isUnknown()) {
-            return ComputationStatus.FAILURE;
-        } else {
+        if (Objects.isNull(angleMonitoringResult) || (!angleMonitoringResult.isDivergent() && !angleMonitoringResult.isUnknown())) {
```
Comment (pet-mit) [inline monitoring/angle-monitoring/src/main/java/com/farao_community/farao/monitoring/angle_monitoring/RaoResultWithAngleMonitoring.java:45]: angleMonitoringResult mustn't be null : maybe require a non null value in constructor, for both objects

**Review by pet-mit — CHANGES_REQUESTED** [review]

File: monitoring/angle-monitoring/src/main/java/com/farao_community/farao/monitoring/angle_monitoring/RaoResultWithAngleMonitoring.java [hunk 3]
```diff
@@ -35,17 +36,23 @@ public RaoResultWithAngleMonitoring(RaoResult raoResult, AngleMonitoringResult a
         this.angleMonitoringResult = angleMonitoringResult;
     }
 
+    public AngleMonitoringResult getAngleMonitoringResult() {
```
[the remark this answers is not in the document]
Comment (MartinBelthle) [reply monitoring/angle-monitoring/src/main/java/com/farao_community/farao/monitoring/angle_monitoring/RaoResultWithAngleMonitoring.java:39]: I used it to know the status of the AngleMonitoringResult (divergent, unsecure etc.) which does not exist in the new class RaoResult. 

Should I put the methods isDivergent() etc. in the new RaoResult class or should we ignore these methods and only use the RaoResult.getComputationStatus() which is not as informative?

[the remark this answers is not in the document]
Comment (pet-mit) [reply monitoring/angle-monitoring/src/main/java/com/farao_community/farao/monitoring/angle_monitoring/RaoResultWithAngleMonitoring.java:39]: yes that was the idea. you'll need to add some logic in the CNE

File: monitoring/angle-monitoring/src/main/java/com/farao_community/farao/monitoring/angle_monitoring/RaoResultWithAngleMonitoring.java (hunk 2 above)
Comment (MartinBelthle) [reply monitoring/angle-monitoring/src/main/java/com/farao_community/farao/monitoring/angle_monitoring/RaoResultWithAngleMonitoring.java]: I added the nullity check for angle monitoring result only

File: data/result-exporter/swe-cne-exporter/src/main/java/com/farao_community/farao/data/swe_cne_exporter/SweRemedialActionSeriesCreator.java [hunk 4]
```diff
@@ -128,30 +130,34 @@ public List<RemedialActionSeries> generateRaSeriesReference(Contingency continge
...
         }
-        if (Objects.nonNull(angleMonitoringResult) && !angleMonitoringResult.isDivergent()) {
+        context.getCreatedIds().stream().sorted()
+                .map(crac::getRemedialAction)
+                .filter(ra -> {
+                    if (raoResult instanceof RaoResultImpl) {
+                        return raoResult.isActivatedDuringState(state, ra);
+                    } else if (raoResult instanceof RaoResultWithAngleMonitoring) {
+                        return ((RaoResultWithAngleMonitoring) raoResult).isActivatedDuringStateButNotByAngleMonitoring(state, ra);
+                    }
```
Comment (pet-mit) [inline data/result-exporter/swe-cne-exporter/src/main/java/com/farao_community/farao/data/swe_cne_exporter/SweRemedialActionSeriesCreator.java:145]: Again, the SWE CNE exporter must not know that the RaoResultWithAngleMonitoring exists.
The "isActivatedDuringState" in RaoResultWithAngleMonitoring now returns RAs activated by RAO or by angle monitoring. So you don't need to separate things anymore (first add RAs from RaoResult, then add them from AngleMonitoring), the code should become simpler than before (only use RAs in RaoReult).

File: data/result-exporter/swe-cne-exporter/src/test/java/com/farao_community/farao/data/swe_cne_exporter/SweAdditionalConstraintSeriesCreatorTest.java [hunk 5]
```diff
@@ -12,8 +12,10 @@
 import com.farao_community.farao.data.crac_api.cnec.AngleCnec;
 import com.farao_community.farao.data.crac_creation.creator.cim.crac_creator.CimCracCreationContext;
 import com.farao_community.farao.data.crac_creation.creator.cim.crac_creator.cnec.AngleCnecCreationContext;
+import com.farao_community.farao.data.rao_result_api.ComputationStatus;
+import com.farao_community.farao.data.rao_result_api.RaoResult;
 import com.farao_community.farao.data.swe_cne_exporter.xsd.AdditionalConstraintSeries;
-import com.farao_community.farao.monitoring.angle_monitoring.AngleMonitoringResult;
+import com.farao_community.farao.monitoring.angle_monitoring.RaoResultWithAngleMonitoring;
```
Comment (pet-mit) [inline data/result-exporter/swe-cne-exporter/src/test/java/com/farao_community/farao/data/swe_cne_exporter/SweAdditionalConstraintSeriesCreatorTest.java:18]: you shouldn't need this

File: data/result-exporter/swe-cne-exporter/src/test/java/com/farao_community/farao/data/swe_cne_exporter/SweRemedialActionSeriesCreatorTest.java [hunk 6]
```diff
@@ -38,12 +40,13 @@ class SweRemedialActionSeriesCreatorTest {
...
     public void setup() {
         this.crac = Mockito.mock(Crac.class);
-        this.raoResult = Mockito.mock(RaoResult.class);
+        this.raoResult = Mockito.mock(RaoResultImpl.class);
         this.cracCreationContext = Mockito.mock(CimCracCreationContext.class);
         this.cneHelper = Mockito.mock(SweCneHelper.class);
 
         Mockito.when(cneHelper.getCrac()).thenReturn(crac);
         Mockito.when(cneHelper.getRaoResult()).thenReturn(raoResult);
+        Mockito.when(raoResult.getComputationStatus()).thenReturn(ComputationStatus.DEFAULT);
```
Comment (pet-mit) [inline data/result-exporter/swe-cne-exporter/src/test/java/com/farao_community/farao/data/swe_cne_exporter/SweRemedialActionSeriesCreatorTest.java:49]: revert this change

File: monitoring/angle-monitoring/src/main/java/com/farao_community/farao/monitoring/angle_monitoring/RaoResultWithAngleMonitoring.java [hunk 7]
```diff
@@ -32,6 +33,9 @@ public class RaoResultWithAngleMonitoring extends RaoResultClone {
     public RaoResultWithAngleMonitoring(RaoResult raoResult, AngleMonitoringResult angleMonitoringResult) {
         super(raoResult);
         this.raoResult = raoResult;
+        if (Objects.isNull(angleMonitoringResult)) {
```
Comment (pet-mit) [inline monitoring/angle-monitoring/src/main/java/com/farao_community/farao/monitoring/angle_monitoring/RaoResultWithAngleMonitoring.java:36]:
```suggestion
        if (angleMonitoringResult == null) {
```

File: monitoring/angle-monitoring/src/main/java/com/farao_community/farao/monitoring/angle_monitoring/RaoResultWithAngleMonitoring.java [hunk 8]
```diff
@@ -32,6 +33,9 @@ public class RaoResultWithAngleMonitoring extends RaoResultClone {
...
+            throw new FaraoException("The AngleMonitoringResult is null");
```
Comment (pet-mit) [inline monitoring/angle-monitoring/src/main/java/com/farao_community/farao/monitoring/angle_monitoring/RaoResultWithAngleMonitoring.java:37]:
```suggestion
            throw new FaraoException("The AngleMonitoringResult must not be null");
```

File: monitoring/angle-monitoring/src/main/java/com/farao_community/farao/monitoring/angle_monitoring/RaoResultWithAngleMonitoring.java [hunk 9]
```diff
@@ -78,4 +82,7 @@ public boolean isActivatedDuringState(State state, NetworkAction networkAction)
         return isActivatedDuringState(state, (RemedialAction<?>) networkAction);
     }
 
+    public boolean isActivatedDuringStateButNotByAngleMonitoring(State state, RemedialAction<?> remedialAction) {
+        return raoResult.isActivatedDuringState(state, remedialAction);
+    }
```
Comment (pet-mit) [inline monitoring/angle-monitoring/src/main/java/com/farao_community/farao/monitoring/angle_monitoring/RaoResultWithAngleMonitoring.java:87]: remove

File: data/result-exporter/swe-cne-exporter/src/main/java/com/farao_community/farao/data/swe_cne_exporter/SweRemedialActionSeriesCreator.java [hunk 10]
```diff
@@ -128,30 +130,34 @@ public List<RemedialActionSeries> generateRaSeriesReference(Contingency continge
...
         RaoResult raoResult = sweCneHelper.getRaoResult();
-        AngleMonitoringResult angleMonitoringResult = sweCneHelper.getAngleMonitoringResult();
         Crac crac = sweCneHelper.getCrac();
         List<RemedialAction<?>> usedRas = new ArrayList<>();
-        if (Objects.nonNull(raoResult)) {
-            context.getCreatedIds().stream().sorted()
-                    .map(crac::getRemedialAction)
-                    .filter(ra -> raoResult.isActivatedDuringState(state, ra))
-                            .forEach(usedRas::add);
+        if (Objects.isNull(raoResult)) {
```
Comment (pet-mit) [inline data/result-exporter/swe-cne-exporter/src/main/java/com/farao_community/farao/data/swe_cne_exporter/SweRemedialActionSeriesCreator.java:135]:
```suggestion
        if (raoResult == null) {
```

**Review by pet-mit — CHANGES_REQUESTED** [review]

File: monitoring/angle-monitoring/src/main/java/com/farao_community/farao/monitoring/angle_monitoring/RaoResultWithAngleMonitoring.java (hunk 7 above)
Comment (pet-mit) [reply monitoring/angle-monitoring/src/main/java/com/farao_community/farao/monitoring/angle_monitoring/RaoResultWithAngleMonitoring.java:36]: also check if raoResult is null, do the same

**Review by pet-mit — APPROVED** [review]
